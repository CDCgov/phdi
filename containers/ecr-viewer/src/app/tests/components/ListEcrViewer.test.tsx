import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import ListECRViewer from "@/app/components/ListEcrViewer";
import userEvent, { UserEvent } from "@testing-library/user-event";

const mockPush = jest.fn();
const mockSearchParams = new URLSearchParams();
jest.mock("next/navigation", () => {
  return {
    useRouter: () => ({
      push: mockPush,
    }),
    useSearchParams: () => mockSearchParams,
    usePathname: () => "",
  };
});

describe("Home Page, ListECRViewer", () => {
  let container: HTMLElement;
  beforeAll(() => {
    container = render(
      <ListECRViewer totalCount={100}>
        <br />
      </ListECRViewer>,
    ).container;
  });
  it("should match snapshot", () => {
    expect(container).toMatchSnapshot();
  });
  it("should pass accessibility test", async () => {
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("Pagination for home page", () => {
  let user: UserEvent;

  beforeEach(() => {
    user = userEvent.setup();
    jest.resetAllMocks();
  });

  it("should have 4 pages when there are 100 and default page length is used", async () => {
    render(
      <ListECRViewer totalCount={100}>
        <br />
      </ListECRViewer>,
    );

    expect(screen.getByText("1"));
    expect(screen.getByText("2"));
    expect(screen.getByText("3"));
    expect(screen.getByText("4"));
    expect(screen.queryByText("5")).not.toBeInTheDocument();
    expect(screen.getByText("Showing 1-25 of 100 eCRs"));
  });

  it("should only update the route once on load", () => {
    render(
      <ListECRViewer totalCount={100}>
        <br />
      </ListECRViewer>,
    );
    expect(mockPush).toHaveBeenCalledExactlyOnceWith("?itemsPerPage=25");
  });

  it("should display 50 per page when items per page is set to 50", async () => {
    jest.spyOn(Storage.prototype, "setItem");

    render(
      <ListECRViewer totalCount={100}>
        <br />
      </ListECRViewer>,
    );
    await user.selectOptions(screen.getByTestId("Select"), ["50"]);

    expect(screen.getByText("1"));
    expect(screen.getByText("2"));
    expect(screen.getByText("Showing 1-50 of 100 eCRs"));
    expect(mockPush).toHaveBeenLastCalledWith("?itemsPerPage=50");
  });

  it("should update local storage when items per page is set to 50", async () => {
    jest.spyOn(Storage.prototype, "setItem");
    render(
      <ListECRViewer totalCount={100}>
        <br />
      </ListECRViewer>,
    );

    await user.selectOptions(screen.getByTestId("Select"), ["50"]);

    expect(localStorage.setItem).toHaveBeenCalledWith(
      "userPreferences",
      JSON.stringify({ itemsPerPage: 50 }),
    );
  });

  it("should load 50 items per page if 50 was previously set", () => {
    const spyLocalStorage = jest.spyOn(Storage.prototype, "getItem");
    spyLocalStorage.mockImplementationOnce(() =>
      JSON.stringify({ itemsPerPage: 50 }),
    );
    render(
      <ListECRViewer totalCount={100}>
        <br />
      </ListECRViewer>,
    );

    expect(screen.getByText("Showing 1-50 of 100 eCRs")).toBeInTheDocument();
    expect(mockPush).toHaveBeenLastCalledWith("?itemsPerPage=50");
  });

  it("should display 51-51 on third page", async () => {
    mockSearchParams.set("page", "3");
    render(
      <ListECRViewer totalCount={51}>
        <br />
      </ListECRViewer>,
    );

    expect(screen.getByText("Showing 51-51 of 51 eCRs")).toBeInTheDocument();
  });
});
