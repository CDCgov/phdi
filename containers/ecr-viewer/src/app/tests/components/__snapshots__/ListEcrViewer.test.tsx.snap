// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page, ListECRViewer should match snapshot 1`] = `
<div>
  <div
    class="main-container"
  >
    <div
      class="homepage-wrapper"
    >
      <table
        class="usa-table usa-table--borderless width-full table-homepage-list"
        data-testid="table"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              Patient
            </th>
            <th
              scope="col"
            >
              Received Date
            </th>
            <th
              scope="col"
            >
              Encounter Date
            </th>
            <th
              scope="col"
            >
              Reportable Condition
            </th>
            <th
              scope="col"
            >
              RCKMS Rule Summary
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <a
                href="/view-data?id=12345"
              >
                John
                 
                Doe
              </a>
              <br />
              DOB: 01/01/1970
            </td>
            <td>
              04/16/2024 9:40 PM UTC
            </td>
            <td>
              04/16/2024 9:40 PM UTC
            </td>
            <td>
              COVID-19
            </td>
            <td>
              Positive
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="/view-data?id=23456"
              >
                Jane
                 
                Doe
              </a>
              <br />
              DOB: 02/01/1955
            </td>
            <td>
              04/16/2024 9:41 PM UTC
            </td>
            <td>
              04/16/2024 9:40 PM UTC
            </td>
            <td>
              COVID-19
            </td>
            <td>
              Positive
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="/view-data?id=34567"
              >
                Dan
                 
                Doe
              </a>
              <br />
              DOB: 12/01/1984
            </td>
            <td>
              04/16/2024 9:42 PM UTC
            </td>
            <td>
              04/16/2024 9:40 PM UTC
            </td>
            <td>
              COVID-19
            </td>
            <td>
              Positive
            </td>
          </tr>
        </tbody>
      </table>
      <nav
        aria-label="Pagination"
        class="usa-pagination"
      >
        <ul
          class="usa-pagination__list"
        >
          <li
            class="usa-pagination__item usa-pagination__page-no"
          >
            <button
              aria-current="page"
              aria-label="Page 1"
              class="usa-button usa-button--unstyled usa-pagination__button usa-current"
              data-testid="pagination-page-number"
              type="button"
            >
              1
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</div>
`;
